feat: Add support of GitHub App

stanleyz opened the pull request:
This is the first part, adding support to the backend.

Additional changes:

1. Add a new ScheduleServiceBase to be extended and shared by ScheduleVcsService and ScheduleGitHubAppTokenService
2. Update the format and import as I touch the code.

fix #1070

stanleyz commented:
@alfespa17 Is the TerrakubeClient library in the below repo? I'll need to add one to query the new GitHub app token.

https://github.com/AzBuilder/terrakube-spring-boot-starter/blob/main/terrakube-client/src/main/java/org/terrakube/client/TerrakubeClient.java

alfespa17 commented:
> @alfespa17 Is the TerrakubeClient library in the below repo? I'll need to add one to query the new GitHub app token.
> 
> https://github.com/AzBuilder/terrakube-spring-boot-starter/blob/main/terrakube-client/src/main/java/org/terrakube/client/TerrakubeClient.java

Yes that is correct that is the terrakube client library that we are using

stanleyz commented:
I've raised the below to the TerrakubeClient, can you please take a look? Thanks.

https://github.com/AzBuilder/terrakube-spring-boot-starter/pull/43

Additionally, I am beaten by GitTagCache.java, in which I can't use TokenService or any other service that has links to JpaRepository to query existing tokens because doing so will cause Spring injection loop, nor can I use Scheduler service to schedule a job to fetch tokens for new Github app installations because GitTagCache is initialised from entity, which doesn't have a reference to quartz scheduler.  Do you mind taking a look to see how we can archieve this? Or otherwise is it possible to remove the usage of GitTagCache from Module.java, or use a event listener for save activity maybe?

alfespa17 commented:
Hello @stanleyzthan you for your help, I will check this next week, I am gaking a break for a few days

stanleyz commented:
Hi @alfespa17 I've made the registry work by adding a column in the module table and a create hook for the Module entity, please take a look and let me know if anything needs to be changed.

However as mentioned above, I am not quite sure about the GitTagCache.java, can we save the versions to the database instead of Redis cache and use normal service and scheduled jobs to update the versions? In this way, we will have a consistent implementation of entities and services. Saying that however there might be many other things I haven't considered, happy to be corrected.

stanleyz commented:
To further support the versions to be saved in the database, it will help if VCS is down like today when GitHub is down globally, in such situations with the current implementation because the versions can not be retrieved you will get an error when downloading modules from the registry, however the source of the module is saved in Terrakube managed storage.

alfespa17 commented on api/src/main/resources/db/changelog/local/changelog-2.22.1-vcs-key-connection-type.xml:
Hello @stanleyz can you add an "update" here to change all previous records in the database with "OAUTH"


        <update tableName="vcs">
            <column name="connection_type" value="OAUTH"/>
        </update>

stanleyz replied:
Sounds good, I've added an `update` directive, please check. Thanks.

alfespa17 commented on api/src/main/resources/db/changelog/local/changelog-2.22.1-module-github-token-id.xml:
```diff
@@ -0,0 +1,14 @@
+        <addColumn tableName="module">
+            <column name="github_app_token_id" type="varchar(36)" />
+        </addColumn>
+        <addForeignKeyConstraint baseTableName="module" baseColumnNames="github_app_token_id"
```
I have a question about this part.

is this going to allow nullable values for modules that are not using a github app @stanleyz right ?

for example, in other scripts we used something like the following:

       <addColumn tableName="module" >
            <column name="ssh_id" type="varchar(36)">
                <constraints nullable="true" foreignKeyName="fk_module_ssh" references="ssh(id)"/>
            </column>
        </addColumn>

stanleyz replied:
The default value of nullable is true for the `addForeignKeyConstraint`, but all good with it explicitly declared. 

I updated the column line, it's a shame that the `addForeignKeyConstraint` doesn't support `nullable` and yet `constraints` in `column` doesn't support cascade settings.

alfespa17 commented:
By the way I created a beta for the change in the library spring boot starter

https://github.com/AzBuilder/terrakube-spring-boot-starter/releases/tag/0.16.0-beta.1

alfespa17 commented:
> Hi @alfespa17 I've raised the below to the TerrakubeClient, can you please take a look? Thanks.
> 
> [AzBuilder/terrakube-spring-boot-starter#43](https://github.com/AzBuilder/terrakube-spring-boot-starter/pull/43)
> 
> Additionally, I am beaten by GitTagCache.java, in which I can't use TokenService or any other service that has links to JpaRepository to query existing tokens because doing so will cause Spring injection loop, nor can I use Scheduler service to schedule a job to fetch tokens for new Github app installations because GitTagCache is initialised from entity, which doesn't have a reference to quartz scheduler. Do you mind taking a look to see how we can archieve this? Or otherwise is it possible to remove the usage of GitTagCache from Module.java, or use a event listener for save activity maybe?

You can access the token value as I did here passing the VCS entity to the GitTagsCache method

https://github.com/AzBuilder/terrakube/blob/517215d271b38483ad468905e72949e4f021249e/api/src/main/java/org/terrakube/api/rs/module/Module.java#L72

https://github.com/AzBuilder/terrakube/blob/517215d271b38483ad468905e72949e4f021249e/api/src/main/java/org/terrakube/api/rs/module/GitTagsCache.java#L140

https://github.com/AzBuilder/terrakube/blob/517215d271b38483ad468905e72949e4f021249e/api/src/main/java/org/terrakube/api/rs/module/GitTagsCache.java#L161

I added the GitTagsCache because we had some performance issue in the past to get the module version as you can see in this [issue](https://github.com/AzBuilder/terrakube/issues/479) and this [pull request](https://github.com/AzBuilder/terrakube/pull/493)

stanleyz commented:
> You can access the token value as I did here passing the VCS entity to the GitTagsCache method

Yes, that's exactly what I am doing in this PR, I am happy that I settled on the same path you are seeing.

> I added the GitTagsCache because we had some performance issue in the past to get the module version as you can see in this https://github.com/AzBuilder/terrakube/issues/479 and this https://github.com/AzBuilder/terrakube/pull/493

Moving that function to a dedicated GitTagCache that you can return in a failure scenario was good. But I am thinking further of using the scheduled job to fetch the versions which will be saved to the database instead of the Redis cache, the shortfall of this is that you won't see a list of versions straight after the module is created successfully in the Terrakube registry, but it's just a delay of several seconds which the user has to wait for anyway because without the content downloaded using the scheduled job the module is not ready for use.

If you don't mind I can try to wrap up a PR for you to take a look at whether the method suits.

alfespa17 commented:
> > You can access the token value as I did here passing the VCS entity to the GitTagsCache method
> 
> Yes, that's exactly what I am doing in this PR, I am happy that I settled on the same path you are seeing.
> 
> > I added the GitTagsCache because we had some performance issue in the past to get the module version as you can see in this #479 and this #493
> 
> Moving that function to a dedicated GitTagCache that you can return in a failure scenario was good. But I am thinking further of using the scheduled job to fetch the versions which will be saved to the database instead of the Redis cache, the shortfall of this is that you won't see a list of versions straight after the module is created successfully in the Terrakube registry, but it's just a delay of several seconds which the user has to wait for anyway because without the content downloaded using the scheduled job the module is not ready for use.
> 
> If you don't mind I can try to wrap up a PR for you to take a look at whether the method suits.

If you can create a different pull request with your proposal I can take a look.

stanleyz commented:
Yes, that's what I meant, will do that later.

stanleyz commented:
@alfespa17 let me know anything else you want me to change. Thanks.

alfespa17 commented:
Hello @stanleyz there is some test that is failing.

https://github.com/AzBuilder/terrakube/blob/b066e2d863dd3e3658eeb9960c230e5943e227ed/api/src/test/java/org/terrakube/api/ModuleTests.java#L26

Can you check ModuleTests.createModuleAsOrgMember the error is the following:



```
java.lang.NullPointerException: Cannot invoke "org.terrakube.api.rs.vcs.Vcs.getConnectionType()" because the return value of "org.terrakube.api.rs.module.Module.getVcs()" is null
	at org.terrakube.api.rs.hooks.module.ModuleManageHook.checkAndCreateGitHubAppToken(ModuleManageHook.java:97) ~[classes/:na]
	at org.terrakube.api.rs.hooks.module.ModuleManageHook.execute(ModuleManageHook.java:54) ~[classes/:na]
	at org.terrakube.api.rs.hooks.module.ModuleManageHook.execute(ModuleManageHook.java:34) ~[classes/:na]
	at com.yahoo.elide.core.lifecycle.LifeCycleHook.execute(LifeCycleHook.java:45) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.core.lifecycle.LifecycleHookInvoker.lambda$onNext$0(LifecycleHookInvoker.java:43) ~[elide-core-7.1.0.jar:na]
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) ~[na:na]
	at com.yahoo.elide.core.lifecycle.LifecycleHookInvoker.onNext(LifecycleHookInvoker.java:42) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.core.RequestScope.lambda$notifySubscribers$4(RequestScope.java:258) ~[elide-core-7.1.0.jar:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[na:na]
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133) ~[na:na]
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[na:na]
	at com.yahoo.elide.core.RequestScope.notifySubscribers(RequestScope.java:257) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.core.RequestScope.runQueuedPreCommitTriggers(RequestScope.java:291) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.jsonapi.JsonApi.handleRequest(JsonApi.java:288) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.jsonapi.JsonApi.post(JsonApi.java:129) ~[elide-core-7.1.0.jar:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:281) ~[spring-core-6.1.11.jar:6.1.11]
	at org.springframework.cloud.context.scope.GenericScope$LockedScopedProxyFactoryBean.invoke(GenericScope.java:482) ~[spring-cloud-context-4.1.2.jar:4.1.2]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.1.11.jar:6.1.11]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768) ~[spring-aop-6.1.11.jar:6.1.11]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:720) ~[spring-aop-6.1.11.jar:6.1.11]
	at com.yahoo.elide.jsonapi.JsonApi$$SpringCGLIB$$0.post(<generated>) ~[elide-core-7.1.0.jar:na]
	at com.yahoo.elide.spring.controllers.JsonApiController$2.call(JsonApiController.java:112) ~[elide-spring-boot-autoconfigure-7.1.0.jar:na]
	at com.yahoo.elide.spring.controllers.JsonApiController$2.call(JsonApiController.java:102) ~[elide-spring-boot-autoconfigure-7.1.0.jar:na]
	at org.springframework.web.context.request.async.WebAsyncManager.lambda$startCallableProcessing$4(WebAsyncManager.java:348) ~[spring-web-6.1.11.jar:6.1.11]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
```

stanleyz commented:
> Hello @stanleyz there is some test that is failing.

Thanks for checking, added null check.

alfespa17 commented:
By the way i still need to do some testing for this one, I will try to do it during this week

stanleyz commented:
> By the way i still need to do some testing for this one, I will try to do it during this week

No problem, thanks @alfespa17 . It's probably easier to use the Terraform provider PR https://github.com/AzBuilder/terraform-provider-terrakube/pull/63 to test, you will need your Github App to have read content and also read/write to webhook , and additionally read/write to commit status if you need the status to be posted. 

Currently I set the endpoint to accept only PKCS8 pem file, so you'd need to convert the RSA pem provided by GitHub to PKCS8, we can change this behaviour, but I saw you use PKCS8 in other places, so it's probably good to keep them consistent.

stanleyz commented:
@alfespa17 I've added UI to create the GitHub App type VCS, please check, let me know anything needs to be changed.

alfespa17 commented:
Thank for adding the UI @stanleyz , now it is easier to test, there are some issues with the UI.

I cant create oauth apps any more I cant select the option.

![image](https://github.com/user-attachments/assets/d341b7dc-fcf3-4187-940e-a37d97fe81dc)

Can you add something for the installation step?? The first time I ran the app it failed because I was missing this step.

![image](https://github.com/user-attachments/assets/882684e7-f85e-43db-b7a2-0cd741e4d3d0)


I got this inside the executor.

```
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - User Home Directory: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - Workspace git clone directory: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/f1a5b769-25c8-446d-9e6e-b87ed9e93738
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - vcsType: GITHUB
[JGit-FileStoreAttributeReader-1] WARN org.eclipse.jgit.util.FS - Thread[JGit-FileStoreAttributeReader-1,5,main]: got smaller file timestamp on / (/.workspace/mark), /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/f1a5b769-25c8-446d-9e6e-b87ed9e93738/.git: 2024-08-29T15:32:02Z < 2024-08-29T15:32:02.017947129Z. Aborting measurement at resolution PT0.982052871S.
[threadPoolTaskExecutor-1] ERROR org.terrakube.executor.service.workspace.SetupWorkspaceImpl - https://github.com/alfespa17/terraform-module-private.git: not authorized
```

The connection failed it was not able to download the code from the repository.

![image](https://github.com/user-attachments/assets/f0dfeeb3-f171-441d-ab4d-4019a59df7be)

alfespa17 commented:
I also got this error if I use the following:

![image](https://github.com/user-attachments/assets/6680ef48-fb54-4eeb-ad86-bd355b389952)

![image](https://github.com/user-attachments/assets/23cd6951-2a7a-4e4f-854e-43144fa8196a)

![image](https://github.com/user-attachments/assets/d8f460e8-0efa-464c-8ba5-0e9e4e7d02b3)

I can see this inside the logs


```
2024-08-29T15:40:04.045Z  INFO 8163 --- [        task-53] o.t.a.r.h.workspace.WorkspaceManageHook  : Workspace mutation hook for workspace simple/adsfqerqwr
2024-08-29T15:40:04.045Z  INFO 8163 --- [        task-53] o.t.a.r.h.workspace.WorkspaceManageHook  : No hook define for this operation
2024-08-29T15:40:04.342Z  INFO 8163 --- [        task-53] o.t.a.p.scheduler.ScheduleServiceBase    : JobId 51b4f21b-a423-4d36-9a56-c5dac0442f6d on GitHubAppToken exists
2024-08-29T15:40:04.343Z  INFO 8163 --- [        task-53] o.t.a.p.scheduler.ScheduleServiceBase    : Delete GitHubAppToken Trigger 51b4f21b-a423-4d36-9a56-c5dac0442f6d
2024-08-29T15:40:04.343Z  INFO 8163 --- [        task-53] o.t.a.p.scheduler.ScheduleServiceBase    : Create GitHubAppToken Trigger DEFAULT.TerrakubeV2_GitHubAppToken_51b4f21b-a423-4d36-9a56-c5dac0442f6d
2024-08-29T15:41:32.312Z  INFO 8163 --- [        task-57] o.t.a.r.h.workspace.WorkspaceManageHook  : Workspace mutation hook for workspace simple/adsfqerqwr
2024-08-29T15:41:32.312Z  INFO 8163 --- [        task-57] o.t.a.r.h.workspace.WorkspaceManageHook  : No hook define for this operation
2024-08-29T15:41:32.609Z  INFO 8163 --- [        task-57] o.t.a.p.scheduler.ScheduleServiceBase    : JobId f39b0aa5-14ad-4a38-99b2-72f600c77883 on GitHubAppToken exists
2024-08-29T15:41:32.609Z  INFO 8163 --- [        task-57] o.t.a.p.scheduler.ScheduleServiceBase    : Delete GitHubAppToken Trigger f39b0aa5-14ad-4a38-99b2-72f600c77883
2024-08-29T15:41:32.610Z  INFO 8163 --- [        task-57] o.t.a.p.scheduler.ScheduleServiceBase    : Create GitHubAppToken Trigger DEFAULT.TerrakubeV2_GitHubAppToken_f39b0aa5-14ad-4a38-99b2-72f600c77883
2024-08-29T15:41:33.537Z  INFO 8163 --- [        task-58] o.t.a.r.h.workspace.WorkspaceManageHook  : Workspace mutation hook for workspace simple/adsfqerqwr
2024-08-29T15:41:33.537Z  INFO 8163 --- [        task-58] o.t.a.r.h.workspace.WorkspaceManageHook  : No hook define for this operation
2024-08-29T15:41:33.804Z  INFO 8163 --- [        task-58] o.t.a.p.scheduler.ScheduleServiceBase    : JobId 697552fb-b286-4f51-a67a-73cc3fa3ec39 on GitHubAppToken exists
2024-08-29T15:41:33.804Z  INFO 8163 --- [        task-58] o.t.a.p.scheduler.ScheduleServiceBase    : Delete GitHubAppToken Trigger 697552fb-b286-4f51-a67a-73cc3fa3ec39
2024-08-29T15:41:33.805Z  INFO 8163 --- [        task-58] o.t.a.p.scheduler.ScheduleServiceBase    : Create GitHubAppToken Trigger DEFAULT.TerrakubeV2_GitHubAppToken_697552fb-b286-4f51-a67a-73cc3fa3ec39
```

alfespa17 commented:
By the way this is the command that I used to converth the private key download from github

openssl pkcs8 -topk8 -inform PEM -outform PEM -nocrypt -in testterrakube.2024-08-29.private-key.pem -out pkcs8.key

stanleyz commented:
> I cant create oauth apps any more I cant select the option.

hmm, that's weird, I can't reproduce this, I have that running fine on Firefox, Google Chrome and Firefox, can you please check again? and maybe check the developer console to see if there are any errors?

![Screenshot 2024-08-31 at 10 28 01 PM](https://github.com/user-attachments/assets/c31ef8d1-c76f-4a5b-8070-459444b713d0)


> Can you add something for the installation step?? The first time I ran the app it failed because I was missing this step.

I have added some information and links to the page, please check.

![Screenshot 2024-08-31 at 10 29 07 PM](https://github.com/user-attachments/assets/0b4ec93f-c382-4b74-8b9a-c70f7c454369)

> I also got this error if I use the following:

Did you grant the "Webhook" (Read/Write) permission to your app? You also need to review your installation if you update the permission after installation. It works fine without issue for me. Judging from the error message above, it's most likely that you either haven't install the app or the "Webhook(Read/Write)" permission hasn't been correctly granted to your selected repo.

alfespa17 commented:
I think I found the issue you need to change this part to 1,2,3 and 4 in addVCS.jsx

```
const githubItems = [
    {
      label: "Github.com (GitHub App)",
      key: "1",
      onClick: () => {
        handleVCSClick("GITHUB", "STANDALONE");
      },
    },
    {
      label: "Github.com (oAuth App)",
      key: "1",
      onClick: () => {
        handleVCSClick("GITHUB");
      },
    },
    {
      label: "GitHub Enterprise (GitHub App)",
      key: "2",
      onClick: () => {
        handleVCSClick("GITHUB_ENTERPRISE", "STANDALONE");
      },
    },
    {
      label: "GitHub Enterprise (oAuth App)",
      key: "2",
      onClick: () => {
        handleVCSClick("GITHUB_ENTERPRISE");
      },
    },
  ];
```

With only 1 and 2 I get the following error:

```
Warning: Duplicated key '1' used in Menu by path [1]
Warning: Duplicated key '2' used in Menu by path [2]
```

![image](https://github.com/user-attachments/assets/07edde5a-1d80-4298-8c68-6269add7edfe)

alfespa17 commented:
I tried with a second repository and I am getting not authorized

```
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.executor.ExecutorJobImpl - Create Job for Organization d9b58bd3-f3fc-4056-a026-1163297e80a8 Workspace 11194d62-c7fa-4b3a-9f1d-d3c64714877d 
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - User Home Directory: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - Workspace git clone directory: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - vcsType: GITHUB
[threadPoolTaskExecutor-1] ERROR org.terrakube.executor.service.workspace.SetupWorkspaceImpl - https://github.com/alfespa17/terraform-azurerm-random.git: not authorized
[threadPoolTaskExecutor-1] ERROR org.terrakube.executor.service.workspace.security.WorkspaceSecurityImpl - Generate Dex Authentication Private Token
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - Executor WorkingDir: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d
[threadPoolTaskExecutor-1] ERROR org.terrakube.executor.service.executor.ExecutorJobImpl - /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d/commitHash.info (No such file or directory)
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - Step list is not empty...
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.executor.ExecutorJobImpl - Execute Plan for Organization d9b58bd3-f3fc-4056-a026-1163297e80a8 Workspace 11194d62-c7fa-4b3a-9f1d-d3c64714877d 
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Terraform Working Directory: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - ***************************************
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - Initializing Terrakube Job 2 Step 54b343c0-bd45-45fd-bb35-1045fc9c6ac0
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - Running Terraform 1.9.5
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - 

***************************************
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - Running Terraform Init: 
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - Generating backend override file for terraform 1.9.5
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - Creating Local Backend File: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d/terrakube_override.tf
[threadPoolTaskExecutor-1] WARN org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Not using any SSH key to download modules
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformClient - Creating terraform downloader using terraform release URL: https://releases.hashicorp.com/terraform/index.json and tofu release URL: https://api.github.com/repos/opentofu/opentofu/releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Initialize TerraformDownloader using custom URL
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - User Home Directory: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create download temp directory: /home/gitpod/.terraform-spring-boot/download/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create terraform directory: /home/gitpod/.terraform-spring-boot/terraform/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - User Home Directory for tofu download: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create tofu download temp directory: /home/gitpod/.terraform-spring-boot/download/tofu/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create tofu directory: /home/gitpod/.terraform-spring-boot/tofu/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading terraform releases list
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Deleting Temp /home/gitpod/.terraform-spring-boot/terraform-10965934911589157462-release
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Found 338 terraform releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading tofu releases list
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Deleting Temp /home/gitpod/.terraform-spring-boot/tofu-15940398047045868823-release
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Found 28 tofu releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading terraform version 1.9.5 architecture amd64 Type Linux
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - terraform_1.9.5_linux_amd64.zip terraform already exists
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - Initializing the backend...
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - 
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - Successfully configured the backend "local"! Terraform will automatically
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - use this backend unless the backend configuration changes.
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - Initializing provider plugins...
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - 
[ForkJoinPool-1-worker-4] INFO org.terrakube.executor.service.logs.LogsConsumer - Terraform has been successfully initialized!
[threadPoolTaskExecutor-1] WARN org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - No commands to run before terraform operation Job 2
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - ***************************************
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - Running Terraform PLAN
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.logs.LogsConsumer - ***************************************
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - Generating backend override file for terraform 1.9.5
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - Creating Local Backend File: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d/terrakube_override.tf
[threadPoolTaskExecutor-1] WARN org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Not using any SSH key to download modules
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformClient - Creating terraform downloader using terraform release URL: https://releases.hashicorp.com/terraform/index.json and tofu release URL: https://api.github.com/repos/opentofu/opentofu/releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Initialize TerraformDownloader using custom URL
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - User Home Directory: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create download temp directory: /home/gitpod/.terraform-spring-boot/download/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create terraform directory: /home/gitpod/.terraform-spring-boot/terraform/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - User Home Directory for tofu download: /home/gitpod
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create tofu download temp directory: /home/gitpod/.terraform-spring-boot/download/tofu/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Validate/Create tofu directory: /home/gitpod/.terraform-spring-boot/tofu/
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading terraform releases list
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Deleting Temp /home/gitpod/.terraform-spring-boot/terraform-15270062497231252012-release
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Found 338 terraform releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading tofu releases list
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Deleting Temp /home/gitpod/.terraform-spring-boot/tofu-1048645508825855676-release
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Found 28 tofu releases
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - Downloading terraform version 1.9.5 architecture amd64 Type Linux
[threadPoolTaskExecutor-1] INFO org.terrakube.terraform.TerraformDownloader - terraform_1.9.5_linux_amd64.zip terraform already exists
[ForkJoinPool-1-worker-6] INFO org.terrakube.executor.service.logs.LogsConsumer - 
[ForkJoinPool-1-worker-6] INFO org.terrakube.executor.service.logs.LogsConsumer - No changes. Your infrastructure matches the configuration.
[ForkJoinPool-1-worker-6] INFO org.terrakube.executor.service.logs.LogsConsumer - 
[ForkJoinPool-1-worker-6] INFO org.terrakube.executor.service.logs.LogsConsumer - Terraform has compared your real infrastructure against your configuration
[ForkJoinPool-1-worker-6] INFO org.terrakube.executor.service.logs.LogsConsumer - and found no differences, so no changes are needed.
[threadPoolTaskExecutor-1] WARN org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Terraform plan Executed Successfully: true Exit Code: 0
[threadPoolTaskExecutor-1] WARN org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - No commands to run after terraform operation Job true
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - terraformStateFile Path: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d/terraformLibrary.tfPlan true
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfstate.local.LocalTerraformStateImpl - Local state file saved to /home/gitpod/.terraform-spring-boot/local/state/d9b58bd3-f3fc-4056-a026-1163297e80a8/11194d62-c7fa-4b3a-9f1d-d3c64714877d/2/54b343c0-bd45-45fd-bb35-1045fc9c6ac0/terraformLibrary.tfPlan
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfoutput.local.LocalTerraformOutputImpl - blobName: /.terraform-spring-boot/local/output/d9b58bd3-f3fc-4056-a026-1163297e80a8/2/54b343c0-bd45-45fd-bb35-1045fc9c6ac0.tfoutput
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.plugin.tfoutput.local.LocalTerraformOutputImpl - Creating Output File: /home/gitpod/.terraform-spring-boot/local/output/d9b58bd3-f3fc-4056-a026-1163297e80a8/2/54b343c0-bd45-45fd-bb35-1045fc9c6ac0.tfoutput
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - Job 2 is still active
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - JobStatus: pending
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - StepId: 54b343c0-bd45-45fd-bb35-1045fc9c6ac0
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - output: 966
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.status.UpdateJobStatusImpl - outputError: 0
```

alfespa17 commented:
I did more debugging. I added some error messages in the executor component.

```
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - Workspace git clone directory: /home/gitpod/.terraform-spring-boot/executor/d9b58bd3-f3fc-4056-a026-1163297e80a8/72ccffdb-a573-455a-938d-a363f3d2bc45
[threadPoolTaskExecutor-1] INFO org.terrakube.executor.service.workspace.SetupWorkspaceImpl - vcsType: GITHUB
org.eclipse.jgit.api.errors.TransportException: https://github.com/alfespa17/terraform-azurerm-random.git: not authorized
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:249)
        at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:319)
        at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:189)
        at org.terrakube.executor.service.workspace.SetupWorkspaceImpl.downloadWorkspace(SetupWorkspaceImpl.java:147)
        at org.terrakube.executor.service.workspace.SetupWorkspaceImpl.prepareWorkspace(SetupWorkspaceImpl.java:60)
        at org.terrakube.executor.service.executor.ExecutorJobImpl.createJob(ExecutorJobImpl.java:40)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:355)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:113)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.eclipse.jgit.errors.TransportException: https://github.com/alfespa17/terraform-azurerm-random.git: not authorized
        at org.eclipse.jgit.transport.TransportHttp.connect(TransportHttp.java:694)
        at org.eclipse.jgit.transport.TransportHttp.openFetch(TransportHttp.java:465)
        at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:153)
        at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:105)
        at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1480)
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:238)
        ... 17 more
```

I tested the endpoint and I see the correct token value

```
https://8080-stanleyz-terrakube-bescm193vr2.ws-us115.gitpod.io/api/v1/organization/d9b58bd3-f3fc-4056-a026-1163297e80a8/vcs/90f31f99-ab0f-4090-8f46-5b4d99965af9/gitHubAppToken
{
    "data": [
        {
            "type": "github_app_token",
            "id": "9a05a8bf-252f-4866-a820-434664f8568f",
            "attributes": {
                "createdBy": "admin@example.com",
                "createdDate": "2024-09-02T22:01Z",
                "installationId": "54294781",
                "owner": "alfespa17",
                "token": "XXXXXX",
                "updatedBy": "admin@example.com",
                "updatedDate": "2024-09-02T22:01Z"
            },
            "relationships": {
                "vcs": {
                    "data": {
                        "type": "vcs",
                        "id": "90f31f99-ab0f-4090-8f46-5b4d99965af9"
                    }
                }
            }
        }
    ]
}
```

Even if I execute it is working in my terminal so I have no idea why I am getting that error
```
git clone https://dummy:MY_TOKEN_FROM_ABOVE@github.com/alfespa17/terraform-azurerm-random.git
```

stanleyz commented:
Thanks @alfespa17 , that's good progress.

> I think I found the issue you need to change this part to 1,2,3 and 4 in addVCS.jsx

I've updated the page and a few other places to make the other warnings disappear. I'm not sure why it worked on my local before this.

> In the other hand I was able to create the workspace with a webhook, I missed the webhook support when I initialy created the github app, now it is working as you can see here:

That's great, I suspect that was the issue.

> When I am running the workspace it is not getting cloned correctly I get this error.

I think I paid much attention to the token creation and added modules and workspace correctly, but not that much for the creation of successful runs, I've pushed more changes, can you please check again? Thanks. I tested on my local, looking good to me.

> git clone https://dummy:MY_TOKEN_FROM_ABOVE@github.com/alfespa17/terraform-azurerm-random.git

for GitHub App token, the user name should be `x-access-token`, so the command should be 

`git clone https://x-access-token:MY_TOKEN_FROM_ABOVE@github.com/alfespa17/terraform-azurerm-random.git`

alfespa17 commented:
I think you got this working @stanleyz  :1st_place_medal: 

![image](https://github.com/user-attachments/assets/3ee52730-9c55-49d4-b180-d6387cb05819)

![image](https://github.com/user-attachments/assets/c215edb6-d072-4863-b4db-bfa909b73de5)

![image](https://github.com/user-attachments/assets/ae72cfe9-44b0-4d8b-a525-410eca3e4ed7)

alfespa17 commented:
I think I can merge this @stanleyz just upload the missing file `/db/changelog/local/changelog-2.22.2-webhook-triggers.xml`

alfespa17 reviewed (approved):
This looks good thank you for your help implementing the github app

stanleyz commented:
That's great, thanks @alfespa17
